### PR TITLE
s2a: Add S2AStub cleanup handler.

### PR DESCRIPTION
**Pull request by rmehta19**

### CODE REVIEW

**Comment by rmehta19** [comment]
@ejona86 , @larry-safran - does this implementation to make sure stub gets closed look ok? Relevant comment thread: https://github.com/grpc/grpc-java/pull/11113#discussion_r1762139544

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java [hunk 1]
```diff
@@ -233,6 +254,9 @@ public void onSuccess(SslContext sslContext) {
                           SharedResourcePool.forResource(GrpcUtil.SHARED_CHANNEL_EXECUTOR))
                       .newHandler(grpcHandler);
 
+              ctx.pipeline().addAfter(ctx.name(), /* name= */ null,
```
Comment (larry-safran) [inline s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:257]: Since you are using the same name to add after for both this and the tls handler, the second one should actually go before this.  You might want to give it a name and then add the other after that name to correct that and make it clear what is desired.

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java [hunk 2]
```diff
@@ -233,6 +254,9 @@ public void onSuccess(SslContext sslContext) {
...
+                  new S2AStubCleanupNegotiationHandler(handler,
+                    grpcHandler.getNegotiationLogger(), s2aStub));
               // Remove the bufferReads handler and delegate the rest of the handshake to the TLS
```
Comment (larry-safran) [inline s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:260]: The comment doesn't seem to match the code since the event gets fired before the bufferReads handler is removed while the comment seems to me to be in the other order.

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java (hunk 1 above)
Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: The intention is to have the order of the handlers be: s2a handler, tls handler, s2a cleanup handler. In https://github.com/grpc/grpc-java/pull/11600/commits/a7fd45f77f1cb93baa6a47d6e845e8cc0fecdd40, I gave names to the tls and cleanup handler; I added the tlsHandler after first, then added the cleanupHandler after the tlsHandler.

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java (hunk 2 above)
Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Done in https://github.com/grpc/grpc-java/pull/11600/commits/a7fd45f77f1cb93baa6a47d6e845e8cc0fecdd40

**Review by larry-safran — APPROVED** [review]

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java [hunk 3]
```diff
@@ -179,6 +180,26 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
...
+        S2AStub s2aStub) {
+      super(next, logger);
+      this.s2aStub = s2aStub;
+    }
+
+    @Override
+    protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
+      s2aStub.close();
+      fireProtocolNegotiationEvent(ctx);
+      ctx.pipeline().remove(this);
```
Comment (ejona86) [inline s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:198]: I see now that `fireProtocolNegotiationEvent()` does this for us (via replace()).

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java [hunk 4]
```diff
@@ -233,9 +254,12 @@ public void onSuccess(SslContext sslContext) {
...
                       .newHandler(grpcHandler);
 
-              // Remove the bufferReads handler and delegate the rest of the handshake to the TLS
-              // handler.
-              ctx.pipeline().addAfter(ctx.name(), /* name= */ null, handler);
+              // Delegate the rest of the handshake to the TLS handler. and remove the 
+              // bufferReads handler.
+              ctx.pipeline().addAfter(ctx.name(), "tlsHandler", handler);
+              ctx.pipeline().addAfter("tlsHandler", "cleanupHandler",
+                  new S2AStubCleanupNegotiationHandler(handler,
```
Comment (ejona86) [inline s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:261]: This is adding the same handler twice. Looks like we can instead pass the cleanup handler to `tls(...).newHandler(cleanupHandler)` and have the tls handler add it to the pipeline for us. The `grpcHandler` should be passed to this cleanup handler.

I'm concerned that no tests failed as the code is now. I expect `s2aStub.close();` is not being called. I don't know how hard it'd be to detect if a leak happens in the test (e.g., by checking that the server stream still exists), but it is worth consideration (but not required).

Comment (larry-safran) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:261]: If stub.close() is called twice, the second one is a no-op.  I believe that the current code runs tls->cleanup->tls which shouldn't exhibit a problem.

Comment (larry-safran) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java:261]: Wouldn't s2aStub.close() be called once, but then the second time the tls handler was run it would reopen it and that instance wouldn't be closed?

Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Thanks for catching this adding of the tls handler twice. Looking at `tls(...).newHandler(...)` it expects a GrpcHttp2ConnectionHandler, so I don't think we can pass in the custom `S2AStubCleanupNegotiationHandler` to it. I added the `S2AStubCleanupNegotiationHandler` via`addLast`, WDYT?

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java (hunk 3 above)
Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Done.

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java (hunk 4 above)
Comment (ejona86) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Hmm... That's a problem. The current code doesn't seem to do anything. (It seems we really need a test to check the lifecycle.)

I'm not immediately seeing any great ways of dealing with this. I'd be interested in how the lifecycle is managed in other languages.

Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: In golang S2A client, we close the stream to S2A after the handshake: https://github.com/google/s2a-go/blob/main/internal/v2/s2av2.go#L152

Comment (ejona86) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Because the handshake is just a method, that does make it easy to know when it is done. The equivalent in Java is `SSLEngine.getHandshakeStatus() == FINISHED`, but that's not at all easy to detect.

The easiest solution seems to be to pass a (optional) `Runnable` ("handshakeCompleteRunnable"?) to `InternalProtocolNegotiators.tls()` and plumb that through `ClientTlsProtocolNegotiator` and `ClientTlsHandler`. Then in `ClientTlsHandler.propagateTlsComplete()` we call the Runnable. We also need to handle the case where the connection is closed before the handshake finishes; seems calling that same runnable when `userEventTriggered0()` calls `fireExceptionCaught()` would be the easier way to manage that.

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java [hunk 5]
```diff
@@ -634,6 +640,9 @@ protected void userEventTriggered0(ChannelHandlerContext ctx, Object evt) throws
                 .withCause(t)
                 .asRuntimeException();
           }
+          if (handshakeCompleteRunnable.isPresent()) {
+            handshakeCompleteRunnable.get().run();
+          }
           ctx.fireExceptionCaught(t);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java:646]: There's one more fireExceptionCaught in this method, up on line 623/629.

**Review by ejona86 — APPROVED** [review]

File: s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java (hunk 4 above)
Comment (rmehta19) [reply s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java]: Done in https://github.com/grpc/grpc-java/pull/11600/commits/1698b54622602a30f9c3654ec8841e8619f7191d. Thanks for pointing out that corner case where the connection gets closed before the end of the handshaker. I added a new integration test to verify stub gets closed. Confirmed (manally) that not setting the `handshakeCompleteRunnable` field results in stub not getting closed.

File: s2a/src/test/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactoryTest.java [hunk 6]
```diff
@@ -158,7 +158,7 @@ public void createProtocolNegotiatorFactory_buildsAnS2AProtocolNegotiatorOnClien
   public void closeProtocolNegotiator_verifyProtocolNegotiatorIsClosedOnClientSide()
       throws Exception {
     InternalProtocolNegotiator.ClientFactory clientFactory =
-        S2AProtocolNegotiatorFactory.createClientFactory(LOCAL_IDENTITY, channelPool);
+        S2AProtocolNegotiatorFactory.createClientFactory(LOCAL_IDENTITY, channelPool, null);
```
Comment (larry-safran) [inline s2a/src/test/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactoryTest.java:161]: Shouldn't this, and probably some of the others, pass in the runnable to close the stub and verify that the stub was also closed?

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java (hunk 5 above)
Comment (rmehta19) [reply netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java:649]: Done in https://github.com/grpc/grpc-java/pull/11600/commits/6cce1561c84905d3435b9e76195dd7b9bb25b93a

File: s2a/src/test/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactoryTest.java (hunk 6 above)
Comment (rmehta19) [reply s2a/src/test/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactoryTest.java:161]: These unit tests are not creating the full pipeline, so if I create the stub here and pass it in, we don't hit code where the tls handler gets inserted (thus we don't get to pass the runnable to `InternalProtocolNegotiators.tls`). I added a new Integration test which verifies the stub gets closed. Does that sound ok?

Comment (larry-safran) [reply s2a/src/test/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactoryTest.java:161]: That sounds fine.
